### DISCUSSION TITLE
How to use ImageMagick to remove vertical scan lines from scanner

### DISCUSSION THREAD

**Discussion by jwc866** [ImageMagick/ImageMagick#6042] [Digital Image Processing] [answered]
Hello:  I'm a long time ImageMagick user and recently trying to find out how to use it remove vertical scan lines from scanned images, from a document scanner.  This is a common problem with document feed scanners where particles can get stuck and cause lines to go through the entire image.  I've tried to look at this but not sure how to do it exactly.

Thanks, in advance.

**Comment by snibgo** [comment]
Morphology might do the trick.

If you want more specific help, you could post an example.

Please also state the IM version and platform.

**Reply by jwc866** [reply]
I started to read up on morphology, that sounds like it would be on the right track.  Here's an example of an image that has the issue, with the line running vertically near the right side.  There are thin lines that run vertically through the scans here and there at random, as dust and stuff gets stuck in the scanner.  I'm looking for a way to run all images through to fix these lines if possible when found.  If there are any examples of how to do this I would be very grateful, but I will keep reading.  Thanks!

![image](https://user-images.githubusercontent.com/16513814/217652127-7683eb4a-c0c3-41be-913e-68d7b82ef735.png)

I am using ImageMagick 6 on Linux (Debian).

**Reply by fmw42** [reply]
Morphology or median filtering would work easily in Imagemagick. You would make a mask for the lines that need repair. Then apply morphology or median filtering to the whole image. Then use the mask to composite the two images such that the line is replaced from the processed image and the rest of the image would come from the input.  The simplest I think would be just do a horizontal kernel median filter that is wider than your line thickness.

You could also use inpainting (See snibgo's hole filling code at http://im.snibgo.com).

**Reply by jwc866** [reply]
> Morphology or median filtering would work easily in Imagemagick. You would make a mask for the lines that need repair. Then apply morphology or median filtering to the whole image. Then use the mask to composite the two images such that the line is replaced from the processed image and the rest of the image would come from the input. The simplest I think would be just do a horizontal kernel median filter that is wider than your line thickness.
> 
> You could also use inpainting (See snibgo's hole filling code at http://im.snibgo.com).

OK got it on the median filtering.  But is there a way to automatically define the mask for each image (the lines can show up at random columns of the image).  I'm unclear on how to use the mask and combine the images (reading the docs but a little lost at the moment).

Thank you!

**Comment by fmw42** [comment]
If the lines are always a given color, you can threshold on that color or color range with -color-threshold.

Unix syntax:

`magick image.suffix \( +clone -statistic median 7x3 \) mask.suffix -composite result.suffix`

Windows syntax:

`magick image.suffix ( +clone -statistic median 7x3 ) mask.suffix -composite result.suffix`

where mask is white where you have lines and need the median and black elsewhere.

See 
https://imagemagick.org/script/command-line-options.php#statistic
https://imagemagick.org/Usage/compose/#compose

Adjust the 7 to larger if your lines are wider than 3 to 5.  You may want to make the 7 larger by 2 to 6 pixels from the width of the line, but always an odd number.

**Comment by snibgo** [comment, marked as answer]
>  But is there a way to automatically define the mask for each image (the lines can show up at random columns of the image). 

What defines the area of the image that is bad? There is an obvious line on the right. It seems to be one column wide, but is actually slightly wider. There is a less obvious problem on the left, just behind the red car.

Naming your input file as "scanline.png":
```
magick ^
  scanLine.png ^
  -set option:WW %%w ^
  -set option:HH %%h ^
  -morphology Convolve 3x1:-0.5,1,-0.5 ^
  -scale "x1^!" ^
  -colorspace Gray ^
  -auto-level ^
  -threshold 90%% ^
  -morphology dilate disk:1.0 ^
  -scale "%%[WW]x%%[HH]^!" ^
  -negate ^
  mask.png

magick ^
  scanLine.png ^
  -write mpr:INP ^
  -morphology Convolve 5x1:0.5,0,0,0,0.5 ^
  ( mpr:INP ^
    mask.png ^
    -alpha off ^
    -compose CopyOpacity -composite ^
  ) ^
  -compose Over -composite ^
  out.png
```
This is Windows BAT syntax. Adjust for bash.

What does it do? It subtracts from each pixel the average of the two pixels on the left and right. The result is light at the bad column, and at each side and also in small areas around wheels etc. Then we scale to "x1" which averages each column, convert to gray, and threshold at 90% of the maximum. The result is light only at the bad column. We dilate ("thicken") it to be 3 pixels wide, and resize to the original size, and negate. The result is white except for a 3-pixel wide column that is black.

The next command replaces every pixel with the average of the pixel two to the left and two to the right, making a sort-of blur image. Then it copies the opacity from the mask to a clone of the input image, and composites that over the sort-of blur.

The commands could be combined into one.

The result fixes the worst of the bad column on the right.

![out](https://user-images.githubusercontent.com/33812368/217682290-d0ea4d08-a199-47fb-b370-e5d80995bb1b.png)

**Reply by jwc866** [reply]
Thank you very much!  This looks like it will work perfectly since I have found that I cannot create the mask based on the color of the line since it varies slightly with each image.
I converted your script back to bash, however I'm stumped on the second scale parameter and how this works using the property WW and HH defined with -set.  I've been reading the docs and confused to the format it uses.
I have:
```
  -set option:WW %w \
  -set option:HH %h \
```
But here, I'm getting a "invalid argument for option `-scale'" when I run the script.
```
 -scale %[WW]x%[HH]\! \
```
Been trying different formatting here without any luck, any input is greatly appreciated.  Thank you again.

**Reply by fmw42** [reply]
In bash, you need to enclose % commands in double quotes. In this case, one set will suffice for the full argument.

` -scale "%[WW]x%[HH]\!" \`


Same for

```
-set option:WW "%w" \
-set option:HH "%h" \
```

**Reply by jwc866** [reply]
Thank you, I have tried however I'm still getting the same error : (invalid argument for option `-scale': %[WW]x%[HH]\! @ error/convert.c/ConvertImageCommand/2627).  Note that I'm still on ImageMagick 6, so I had to use convert instead of magick but assuming that's not related.  Here's the entire script:

```
#!/bin/bash
convert \
  input.png \
  -set option:WW "%w" \
  -set option:HH "%h" \
  -morphology Convolve 3x1:-0.5,1,-0.5 \
  -scale x1\! \
  -colorspace Gray \
  -auto-level \
  -threshold 90% \
  -morphology dilate disk:1.0 \
  -scale "%[WW]x%[HH]\!" \
  -negate \
  mask.png

convert \
  input.png \
  -write mpr:INP \
  -morphology Convolve 5x1:0.5,0,0,0,0.5 \
  \( mpr:INP \
    mask.png \
    -alpha off \
    -compose CopyOpacity -composite \
  \) \
  -compose Over -composite \
  out.png
```

**Reply by fmw42** [reply]
Those inline computations such as `-scale "%[WW]x%[HH]\!"` only work for IM 7 when using magick in place of convert.  Are you on IM 7? If so, change convert to magick and try again.

**Reply by jwc866** [reply]
I'm on IM 6, stock with Debian 11.  Luckily, I can simply define the WxH for now since all of the images are the same size.  Also, our images are slightly larger than the sample so I had to tweak slightly the second convolve to 7x1.  This is the script I'm using that is working:

```
#!/bin/bash
convert \
  input.png \
  -morphology Convolve 3x1:-0.5,1,-0.5 \
  -scale x1\! \
  -colorspace Gray \
  -auto-level \
  -threshold 90% \
  -morphology dilate disk:1.0 \
  -scale 2025x3000\! \
  -negate \
  mask.png
convert \
  input.png \
  -write mpr:INP \
  -morphology Convolve 7x1:0.5,0,0,0,0,0,0.5 \
  \( mpr:INP \
    mask.png \
    -alpha off \
    -compose CopyOpacity -composite \
  \) \
  -compose Over -composite \
  out.png
```
Thank you very much for your time and help.

**Reply by milahu** [reply]
a slightly extended version of that bash script

<details>

```sh
#!/usr/bin/env bash

# https://github.com/ImageMagick/ImageMagick/discussions/6042#discussioncomment-4918533

[ -e "$1" ] || {
  echo "usage: $0 input.png"
  exit 1
}

base=${1%.*}
ext=${1##*.}

output_path="$base.fixed.$ext"

[ -e "$output_path" ] && {
  echo "error: output file exists: $output_path"
  exit 1
}

touch "$output_path" || {
  echo "error: failed to write output file: $output_path"
  exit 1
}
rm "$output_path"

# TODO is pnm faster than png?
mask_path=/run/user/$(id -u)/mask.pnm



echo "writing $output_path"

convert \
  "$1" \
  -morphology Convolve 3x1:-0.5,1,-0.5 \
  -scale x1\! \
  -colorspace Gray \
  -auto-level \
  -threshold 90% \
  -morphology dilate disk:1.0 \
  -scale 2025x3000\! \
  -negate \
  "$mask_path"

convert \
  "$1" \
  -write mpr:INP \
  -morphology Convolve 7x1:0.5,0,0,0,0,0,0.5 \
  \( mpr:INP \
    "$mask_path" \
    -alpha off \
    -compose CopyOpacity -composite \
  \) \
  -compose Over -composite \
  "$output_path"

rm "$mask_path"
```

</details>


but the script did not work in my case
to remove vertical grey lines from scanned images from an ADF document scanner

at 300dpi, the grey lines are 7px wide, and the lines are noisy

![scan 2023-10-20 18-57-53 1 crop-grey-line](https://github.com/ImageMagick/ImageMagick/assets/12958815/cff55333-332d-4d6b-895b-98ea78c9ab2f)

to remove the grey lines, im using color levels

```
convert input.png -level 0x85% output.png
```

im using gimp to find the level values: gimp -> colors -> levels

edit: nope. to remove all grey lines, i need `-level 0x66%`
but that is too lossy because my input also contains grey lines

edit: fixed by properly cleaning the scanner glass with acetone
probably there was some oily dirt causing these grey lines
